clang/lib/Sema/SemaTemplateInstantiateDecl.cpp

```diff
@@ -2146,9 +2147,15 @@ Decl *TemplateDeclInstantiator::VisitFunctionDecl(
     FunctionDecl *SpecFunc
       = FunctionTemplate->findSpecialization(Innermost, InsertPos);
 
-    // If we already have a function template specialization, return it.
-    if (SpecFunc)
-      return SpecFunc;
+    if (SpecFunc) {
+      if (!SpecFunc->isTemplateInstantiation())
+        return SpecFunc;
+
+      for (auto *Redecl : SpecFunc->redecls())
+        if (Redecl->getPrimaryTemplate() == FunctionTemplate)
+          return Redecl;
+    }
+    PrevFunc = SpecFunc;
```

Comment:
Let me rephrase the purpose here:

So now clang instantiates a new function declaration as long as the previous instantiation doesn't exist / the previous instantiation has a distinct primary template declaration, right?

I feel I struggle to understand the benefit here, though it might indeed have fixed/improved something. Can you explain it more?

Reply:
This fixes the case:
```C++
  cica(A<0>{}, 0);
  (void)A<1>{};
  cica(A<0>{}, 0);
```

As explained in the previous comment.